test/qpy_compat/process_version.sh

```diff
@@ -14,9 +14,23 @@
 set -e
 set -x
 
+# version is the source version, this is the release with which to generate
+# qpy files with to load with the version under test
 version=$1
 parts=( ${version//./ } )
-if [[ ${parts[1]} -lt 18 ]] ; then
+# qiskit_version is the version under test, We're testing that we can correctly
+# read the qpy files generated with source version with this version.
+qiskit_version=`../qiskit_venv/bin/python -c "import qiskit;print(qiskit.__version__)"`
+qiskit_version_pars = ( ${qiskit_version//./ } )
+
+# If source version is less than 0.18 QPY didn't exist yet so exit fast
+if [[ ${parts[0]} -eq 0 && ${parts[1]} -lt 18 ]] ; then
+    exit 0
+fi
+# If the source version is newer than the version under test exit fast because
+# there is no QPY compatibility for loading a qpy file generated from a newer
+# release with an older release of Qiskit.
+if [[ ${parts[0]} -gt ${qiskit_parts[0]} || ${parts[1]} -gt ${qiskit_parts[1]} ]] ; then
```

Comment:
This doesn't implement tuple comparison quite right - the right-hand side of the `||` needs to limit to the case that the minor versions are _equal_ (i.e. forbid `${parts[0]} -lt ${qiskit_parts[1]}`).

If we don't want to get into the weeds on doing that properly in bash, we could do something like

```bash
if [[ $(printf '%s\n%s\n' "$version" "$qiskit_version" | sort -V | head -n1) == $version]]
```

(or whatever it is) and let `sort -V` do the ordering for us.

Reply:
So I played with this locally and the only issue is if it's a pre-release `sort -V` doesn't handle this correctly it treats the suffix as a higher version. So if we're doing an upgrade test from 0.45.0rc1 to 0.45.0 it will evaluate that as trying to use a newer version here. I think I'm just going to write a small python script to do the comparison exactly as we want, rather than try to figure out how to best do this in bash.

Reply:
Oh yeah, now I remember hitting that in #10610, but it didn't matter there because prereleases would never be deployed, so I'd already filtered them out.

Reply:
Done in: https://github.com/Qiskit/qiskit/pull/11101/commits/1c87c502c228942723fd7299af648881687de35e